File: .aoc/languages/Python.py
from .Language import Language

class Python(Language):

    def getFormattedName(self):
        return "[dodger_blue1]Python[/dodger_blue1]"

    def getTemplateDirectory(self):
        return ".aoc/file_templates/python/"

    def getRunCommand(self, day, task, sourcePath):
        return f"python -u {sourcePath}/runner.py {str(task)}"

    def getTestCommand(self, day, task, sourcePath):
        return f"python -u {sourcePath}/runner.py {str(task)} test"

    def hasIndividualTaskRunCommands(self):
        return True

File: runner/languages/python.go
package languages

import (
	"aoc-cli/utils"
	_ "embed"

	"github.com/spf13/viper"
)

type Python struct{}

//go:embed python/runner.py
var pythonRunnerFile string

//go:embed python/task.py
var pythonTaskFile string

func (p Python) getPythonExecutable() string {
	configExecutable := viper.GetString("languages.python.executable")
	if configExecutable == "" {
		return "python"
	} else {
		return configExecutable
	}
}

func (p Python) GetSolveCommand(executionDirectory string, task int) utils.ExecutionDetails {
	return *utils.ToExecute(p.getPythonExecutable()).Arg("-u").Arg("runner.py").Argf("%d", task)
}

func (p Python) GetTestCommand(executionDirectory string, task int) utils.ExecutionDetails {
	return *utils.ToExecute(p.getPythonExecutable()).Arg("-u").Arg("runner.py").Argf("%d", task).Arg("test")
}

func (p Python) GetPreparationCommand(executionDirectory string, task int) []utils.ExecutionDetails {
	return []utils.ExecutionDetails{}
}

func (p Python) GetFilesToWrite() []utils.FileTemplate {
	runnerFile := utils.FileTemplate{Content: pythonRunnerFile, Filename: "runner.py"}
	taskFile := utils.FileTemplate{Content: pythonTaskFile, Filename: "task.py"}

	return []utils.FileTemplate{runnerFile, taskFile}
}

func (p Python) GetLanguageSpecificConfigKeys() map[string]utils.FlagMetadata {
	return map[string]utils.FlagMetadata{
		"executable": {Description: "Python executable to run", DataType: "string", ViperKey: "executable"},
	}
}
